llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -1983,6 +1983,41 @@ void CombinerHelper::applyShiftOfShiftedLogic(MachineInstr &MI,
   MI.eraseFromParent();
 }
 
+bool CombinerHelper::matchLsbClearByShifts(MachineInstr &MI,
+                                           BuildFnTy &MatchInfo) {
+  // fold (A >> C) << C to A & K, where K = (-1 >> C) << C
+  const GShl *Shl = cast<GShl>(&MI);
+  GLShr *Lshr = cast<GLShr>(MRI.getVRegDef(Shl->getSrcReg()));
+
+  if (!MRI.hasOneNonDBGUse(Lshr->getReg(0)))
+    return false;
+
+  APInt C1, C2;
+  if (!mi_match(Shl->getShiftReg(), MRI, m_ICstOrSplat(C1)) ||
+      !mi_match(Lshr->getShiftReg(), MRI, m_ICstOrSplat(C2)))
+    return false;
```

Comment:
This was already matched in the pattern? Ideally we could directly pass in the constant, but failing that should just assert on the failure

Reply:
No, this matching the constant operands, which was not matched in the pattern.

Reply:
You can use `getIConstantFromReg`.

Reply:
With the current implementation I think I cannot use `getIConstantFromReg` since that expects that the operand is G_CONSTANT, but it is not guaranteed. If in the end G_CONSTANT will be used in the pattern, then indeed it usable and will change it to that.